File: Cascade/Core/PS2/DMAC.swift
import Foundation

// MARK: - EE DMA Controller
// 10 DMA channels for bulk transfers between RAM, GIF, VIF0/1, SPR, SIF0/1/2, IPU, etc.

public final class DMAC {

    static let channelCount = 10

    struct Channel {
        var madr: UInt32 = 0    // memory address
        var qwc:  UInt32 = 0    // quadword count
        var tadr: UInt32 = 0    // tag address
        var chcr: UInt32 = 0    // channel control
        var sadr: UInt32 = 0    // scratchpad address

        var active: Bool { chcr & (1 << 8) != 0 }
    }

    var channels: [Channel] = Array(repeating: Channel(), count: DMAC.channelCount)
    var dstat: UInt32 = 0    // interrupt status
    var dmask: UInt32 = 0    // interrupt mask
    var ctrl:  UInt32 = 0    // global control
    var sqwc:  UInt32 = 0    // stall quad-word count
    var rbsr:  UInt32 = 0    // ring buffer size
    var rbor:  UInt32 = 0    // ring buffer offset

    weak var bus: MemoryBus?
    weak var intc: INTC?
    weak var gs: GraphicsSynthesizer?
    weak var vif0: VectorInterface?
    weak var vif1: VectorInterface?

    func step() {
        for i in 0..<DMAC.channelCount {
            guard channels[i].active else { continue }
            performTransfer(channel: i)
        }
    }

    private func performTransfer(channel idx: Int) {
        var ch = channels[idx]
        guard ch.qwc > 0, let bus = bus else {
            channels[idx].chcr &= ~(1 << 8)   // clear STR
            dstat |= (1 << idx)
            checkInterrupt()
            return
        }

        let batchSize = min(ch.qwc, 16)
        switch idx {
        case 0: // VIF0 — feed data into VIF0
            if let vif = vif0 {
                var words: [UInt32] = []
                for _ in 0..<(batchSize * 4) {
                    words.append(bus.read32(address: ch.madr))
                    ch.madr += 4
                }
                ch.qwc -= batchSize
                vif.feed(words: words)
            } else {
                ch.madr += batchSize * 16
                ch.qwc  -= batchSize
            }
        case 1: // VIF1 — feed data into VIF1
            if let vif = vif1 {
                var words: [UInt32] = []
                for _ in 0..<(batchSize * 4) {
                    words.append(bus.read32(address: ch.madr))
                    ch.madr += 4
                }
                ch.qwc -= batchSize
                vif.feed(words: words)
            } else {
                ch.madr += batchSize * 16
                ch.qwc  -= batchSize
            }
        case 2: // GIF — direct path to GS
            var gifData: [UInt64] = []
            for _ in 0..<batchSize {
                let lo = UInt64(bus.read32(address: ch.madr))
                let hi = UInt64(bus.read32(address: ch.madr + 4))
                gifData.append(lo | (hi << 32))
                ch.madr += 16
                ch.qwc  -= 1
            }
            gs?.processGIFPacket(data: gifData, qwordCount: gifData.count, flag: 0)
        default:
            ch.madr += batchSize * 16
            ch.qwc  -= batchSize
        }

        channels[idx] = ch

        if ch.qwc == 0 {
            channels[idx].chcr &= ~(1 << 8)
            dstat |= (1 << idx)
            checkInterrupt()
        }
    }

    private func checkInterrupt() {
        if dstat & dmask != 0 { intc?.assertIRQ(bit: 3) }
    }

    // MARK: - I/O

    func read(offset: UInt32) -> UInt32 {
        let ch = Int((offset >> 8) & 0xF)
        switch offset & 0xFF {
        case 0x00: return ch < channels.count ? channels[ch].chcr : 0
        case 0x10: return ch < channels.count ? channels[ch].madr : 0
        case 0x20: return ch < channels.count ? channels[ch].qwc  : 0
        case 0x30: return ch < channels.count ? channels[ch].tadr : 0
        case 0xE0: return dstat
        case 0xF0: return dmask
        default:   return 0
        }
    }

    func write(offset: UInt32, value: UInt32) {
        let ch = Int((offset >> 8) & 0xF)
        switch offset & 0xFF {
        case 0x00:
            if ch < channels.count {
                channels[ch].chcr = value
                if value & (1 << 8) != 0 { performTransfer(channel: ch) }
            }
        case 0x10: if ch < channels.count { channels[ch].madr = value }
        case 0x20: if ch < channels.count { channels[ch].qwc  = value }
        case 0x30: if ch < channels.count { channels[ch].tadr = value }
        case 0xE0: dstat = value
        case 0xF0: dmask = value
        default:   break
        }
    }
}

// MARK: - INTC

public final class INTC {
    var stat: UInt32 = 0
    var mask: UInt32 = 0

    func assertIRQ(bit: Int) {
        stat |= (1 << bit)
    }

    func read(offset: UInt32) -> UInt32 {
        switch offset {
        case 0: return stat
        case 4: return mask
        default: return 0
        }
    }

    func write(offset: UInt32, value: UInt32) {
        switch offset {
        case 0: stat &= ~value
        case 4: mask = value
        default: break
        }
    }
}

// MARK: - EE Timer

public final class EETimer {
    var count: [UInt32] = [0, 0, 0, 0]
    var mode:  [UInt32] = [0, 0, 0, 0]
    var comp:  [UInt32] = [0, 0, 0, 0]
    weak var intc: INTC?

    func tick() {
        for i in 0..<4 {
            guard mode[i] & 1 != 0 else { continue }
            count[i] &+= 1
            if count[i] >= comp[i] {
                count[i] = 0
                intc?.assertIRQ(bit: 9 + i)
            }
        }
    }

    func read(offset: UInt32) -> UInt32 {
        let t = Int((offset >> 4) & 3)
        switch offset & 0xF {
        case 0: return count[t]
        case 4: return mode[t]
        case 8: return comp[t]
        default: return 0
        }
    }

    func write(offset: UInt32, value: UInt32) {
        let t = Int((offset >> 4) & 3)
        switch offset & 0xF {
        case 0: count[t] = value
        case 4: mode[t]  = value
        case 8: comp[t]  = value
        default: break
        }
    }
}

File: Cascade/Core/CascadeCore/dmac.cpp
#include "dmac.h"
#include "bus.h"
#include "gs.h"
#include "vu.h"
#include "intc.h"
#include "spu2.h"
#include "iop.h"
#include <cstring>
#include <algorithm>

void DMAC::reset() {
    memset(ch, 0, sizeof(ch));
    ctrl   = 0;
    stat   = 0;
    pcr    = 0;
    sqwc   = 0;
    rbsr   = 0;
    rbor   = 0;
    stadr  = 0;
    enableR = enableW = 0;
}

// ── Read/Write ────────────────────────────────────────────────────────────────
// Channel layout (relative to 0x8000 base):
//   VIF0=0x0000, VIF1=0x1000, GIF=0x2000, IPUFROM=0x3000, IPUTO=0x4000
//   SIF0=0x5000, SIF1=0x6000, SIF2=0x7000, SPRFROM=0x8000, SPRTO=0x9000
//
// Within each channel (4-byte stride):
//   +0x00 = CHCR, +0x10 = MADR, +0x20 = QWC, +0x30 = TADR, +0x40 = ASR0, +0x50 = ASR1

static int channelFromOffset(u32 off, u32& regOff) {
    // Try channel-sized slots at 0x8000 + id*0x1000
    if (off >= 0x8000 && off < 0x9000) { regOff = off & 0xFF; return DMA_VIF0; }
    if (off >= 0x9000 && off < 0xA000) { regOff = off & 0xFF; return DMA_VIF1; }
    if (off >= 0xA000 && off < 0xB000) { regOff = off & 0xFF; return DMA_GIF;  }
    if (off >= 0xB000 && off < 0xB400) { regOff = off & 0xFF; return DMA_IPUFROM; }
    if (off >= 0xB400 && off < 0xC000) { regOff = off & 0xFF; return DMA_IPUTO;   }
    if (off >= 0xC000 && off < 0xC800) { regOff = off & 0xFF; return DMA_SIF0;    }
    if (off >= 0xC800 && off < 0xD000) { regOff = off & 0xFF; return DMA_SIF1;    }
    if (off >= 0xD000 && off < 0xD400) { regOff = off & 0xFF; return DMA_SIF2;    }
    if (off >= 0xD400 && off < 0xD800) { regOff = off & 0xFF; return DMA_SPRFROM; }
    if (off >= 0xD800 && off < 0xE000) { regOff = off & 0xFF; return DMA_SPRTO;   }
    regOff = off;
    return -1;
}

u32 DMAC::read(u32 offset) {
    u32 regOff = 0;
    int id = channelFromOffset(offset, regOff);
    if (id >= 0 && id < DMA_CHAN_COUNT) {
        DmaChannel& c = ch[id];
        switch (regOff & ~3u) {
        case 0x00: return c.chcr;
        case 0x10: return c.madr;
        case 0x20: return c.qwc;
        case 0x30: return c.tadr;
        case 0x40: return c.asr[0];
        case 0x50: return c.asr[1];
        default:   return 0;
        }
    }

    // Global DMAC registers (relative to 0xE000 or passed raw)
    u32 gOff = (offset >= 0xE000) ? offset : offset;
    switch (gOff) {
    case 0xE000: return ctrl;
    case 0xE010: return stat;
    case 0xE020: return pcr;
    case 0xE030: return sqwc;
    case 0xE040: return rbsr;
    case 0xE050: return rbor;
    case 0xE060: return stadr;
    case 0xF520: return enableR; // D_ENABLER
    case 0xF590: return enableW; // D_ENABLEW
    default:     return 0;
    }
}

void DMAC::write(u32 offset, u32 value) {
    u32 regOff = 0;
    int id = channelFromOffset(offset, regOff);
    if (id >= 0 && id < DMA_CHAN_COUNT) {
        DmaChannel& c = ch[id];
        switch (regOff & ~3u) {
        case 0x00: c.chcr = value; break;
        case 0x10: c.madr = value & 0x1FFF'FFF0u; break; // QW-aligned
        case 0x20: c.qwc  = value & 0xFFFFu; break;
        case 0x30: c.tadr = value & 0x1FFF'FFF0u; break;
        case 0x40: c.asr[0] = value; break;
        case 0x50: c.asr[1] = value; break;
        default:   break;
        }
        return;
    }

    u32 gOff = offset;
    switch (gOff) {
    case 0xE000: ctrl  = value; break;
    case 0xE010:
        // Writing 1 to a stat bit clears it (acknowledge)
        stat &= ~(value & 0xFFFFu);
        // Upper 16 bits control DMAC channel enable mask
        stat = (stat & 0xFFFFu) | (value & 0xFFFF'0000u);
        break;
    case 0xE020: pcr   = value; break;
    case 0xE030: sqwc  = value; break;
    case 0xE040: rbsr  = value; break;
    case 0xE050: rbor  = value; break;
    case 0xE060: stadr = value; break;
    case 0xF520: enableR = value; break;
    case 0xF590: enableW = value; break;
    default: break;
    }
}

// ── Main step ─────────────────────────────────────────────────────────────────

void DMAC::step() {
    if (!(ctrl & 1)) return; // DMA master enable

    for (int id = 0; id < DMA_CHAN_COUNT; id++) {
        if (ch[id].active()) {
            runChannel(id);
        }
    }
}

void DMAC::runChannel(int id) {
    switch (id) {
    case DMA_GIF:     runGIF (ch[DMA_GIF]);     break;
    case DMA_VIF1:    runVIF1(ch[DMA_VIF1]);    break;
    case DMA_SIF0:    runSIF0(ch[DMA_SIF0]);    break;
    case DMA_SIF1:    runSIF1(ch[DMA_SIF1]);    break;
    case DMA_IPUFROM: finishTransfer(id);        break; // IPU → EE: stub drain
    default: {
        // Generic normal-mode: drain QWC
        DmaChannel& c = ch[id];
        u32 drain = std::min(c.qwc, 16u);
        c.madr += drain * 16;
        c.qwc  -= drain;
        if (c.qwc == 0) finishTransfer(id);
        break;
    }
    }
}

// ── GIF DMA (channel 2: EE RAM → GS via GIF) ─────────────────────────────────

void DMAC::runGIF(DmaChannel& c) {
    if (!gs || !bus) { c.chcr &= ~0x100u; return; }

    static constexpr int MAX_QW_PER_STEP = 64;

    int mod = c.mod();
    if (mod == 1) { // Chain mode
        if (c.qwc == 0) {
            // Fetch next tag from TADR
            if (!bus) { finishTransfer(DMA_GIF); return; }
            u64 tag  = bus->read64(c.tadr);
            u32 qwc  = (u32)(tag & 0x7FFF);
            u32 id2  = (u32)((tag >> 28) & 7);
            bool irq = (tag >> 31) & 1;
            c.qwc    = qwc;
            // Optionally propagate tag fields to CHCR[31:16]
            c.chcr = (c.chcr & 0x0000'FFFFu) | (u32)((tag >> 16) & 0xFFFF) << 16;

            switch (id2) {
            case 0: // refe — transfer from addr in tag, then stop
                c.madr = (u32)((tag >> 32) & 0x1FFF'FFF0u);
                // Don't update TADR; will stop after transfer
                break;
            case 1: // cnt — transfer from MADR, TADR = MADR+qwc*16 after
                c.madr = c.tadr + 16;
                c.tadr = c.madr + qwc * 16;
                break;
            case 2: // next — transfer from MADR, TADR = addr in tag
                c.madr = c.tadr + 16;
                c.tadr = (u32)((tag >> 32) & 0x1FFF'FFF0u);
                break;
            case 3: // ref — transfer from addr in tag, TADR advances past tag
                c.madr = (u32)((tag >> 32) & 0x1FFF'FFF0u);
                c.tadr += 16;
                break;
            case 4: // refs — same as ref for us
                c.madr = (u32)((tag >> 32) & 0x1FFF'FFF0u);
                c.tadr += 16;
                break;
            case 7: // end — transfer from MADR, stop after
                c.madr = c.tadr + 16;
                c.tadr = 0;
                c.chcr &= ~0x100u; // pre-mark done; we'll really stop after transfer
                break;
            default:
                finishTransfer(DMA_GIF);
                return;
            }

            if (irq) {
                finishTransfer(DMA_GIF);
                return;
            }
        }
    }

    if (c.qwc == 0) { finishTransfer(DMA_GIF); return; }

    int batch = std::min((int)c.qwc, MAX_QW_PER_STEP);

    // Read QWs from EE RAM and feed to GS
    static u64 qwbuf[MAX_QW_PER_STEP * 2];
    for (int i = 0; i < batch * 2; i++)
        qwbuf[i] = bus->read64(c.madr + (u32)(i * 8));
    gs->processGIF(qwbuf, batch * 2);

    c.madr += (u32)(batch * 16);
    c.qwc  -= (u32)batch;

    if (c.qwc == 0 && mod != 1) {
        finishTransfer(DMA_GIF);
    }
    // In chain mode, we loop back next step to fetch new tag
}

// ── VIF1 DMA (channel 1: EE RAM → VU1 via VIF1) ──────────────────────────────

void DMAC::runVIF1(DmaChannel& c) {
    if (!bus) { finishTransfer(DMA_VIF1); return; }
    // VIF1 would decode VIF opcodes; for now just drain
    u32 drain = std::min(c.qwc, 16u);
    c.madr += drain * 16;
    c.qwc  -= drain;
    if (c.qwc == 0) finishTransfer(DMA_VIF1);
}

// ── SIF0 DMA (IOP→EE) ────────────────────────────────────────────────────────

void DMAC::runSIF0(DmaChannel& c) {
    if (!iop || !bus) { finishTransfer(DMA_SIF0); return; }
    // Copy a 32-byte chunk from IOP RAM to EE RAM
    if (c.qwc == 0) { finishTransfer(DMA_SIF0); return; }

    u32 drain = std::min(c.qwc, 4u); // 4 QW per step
    for (u32 i = 0; i < drain; i++) {
        // Read 16 bytes from IOP RAM via SADR
        u32 iopOff = c.sadr + i * 16;
        for (int b = 0; b < 16; b++) {
            u8 byte = (iopOff + b < IOP_RAM_SIZE && iop->ram) ? iop->ram[iopOff + b] : 0;
            bus->write8(c.madr + i * 16 + b, byte);
        }
    }
    c.sadr += drain * 16;
    c.madr += drain * 16;
    c.qwc  -= drain;
    if (c.qwc == 0) finishTransfer(DMA_SIF0);
}

// ── SIF1 DMA (EE→IOP) ────────────────────────────────────────────────────────

void DMAC::runSIF1(DmaChannel& c) {
    if (!iop || !bus) { finishTransfer(DMA_SIF1); return; }
    if (c.qwc == 0) { finishTransfer(DMA_SIF1); return; }

    u32 drain = std::min(c.qwc, 4u);
    for (u32 i = 0; i < drain; i++) {
        u32 iopOff = c.sadr + i * 16;
        for (int b = 0; b < 16; b++) {
            u8 byte = bus->read8(c.madr + i * 16 + b);
            if (iopOff + b < IOP_RAM_SIZE && iop->ram)
                iop->ram[iopOff + b] = byte;
        }
    }
    c.sadr += drain * 16;
    c.madr += drain * 16;
    c.qwc  -= drain;
    if (c.qwc == 0) finishTransfer(DMA_SIF1);
}

// ── Finish transfer ────────────────────────────────────────────────────────────

void DMAC::finishTransfer(int id) {
    DmaChannel& c = ch[id];
    c.chcr &= ~0x100u; // clear STR bit
    c.qwc   = 0;

    // Set DMAC_STAT bit for this channel and raise INTC IRQ if enabled
    stat |= (1u << id);

    // DMAC raises INTC bit 3 (DMA) when any channel completes and its
    // PCR channel-priority bit is set for interrupt generation.
    // Simplified: always raise INTC DMAC interrupt (bit 3 = VBLANK_E conflicts;
    // use bit 8+id range for DMA bits 8–17 per DMAC_STAT layout).
    // INTC bit for DMAC channel id: bits 8–17 in DMAC_STAT map to INTC?
    // Actually, DMAC channels generate an INTC bit 0 (GS) or route through DMAC_STAT.
    // On real hardware, the DMAC IRQ fires into INTC bit 3 (DMAC).
    if (intc) {
        // Check if this channel's interrupt mask bit in DMAC STAT is enabled
        bool chEnabled = (stat >> (16 + id)) & 1;
        if (!chEnabled) {
            // Channel interrupt mask not set → raise anyway for correctness
        }
        // Raise INTC DMAC interrupt line (bit 3 on EE)
        intc->assertIRQ(3);
    }
}
